Allow the usage of React components as UI extensions

martin-fleck-at opened the pull request:
- Provide dedicated package for editor UI
- Allow usage of VNode and React in same package through global.d.ts
   - Correct typing defined by annotation (@jsx svg) or default (React)

ivy-lli commented on packages/inscription/src/inscription/inscription-ui.tsx:
```diff
@@ -55,38 +54,45 @@ export class InscriptionUi extends GLSPAbstractUIExtension implements IActionHan
+      return;
+    }
+    return (
+      <React.StrictMode>
```
I think this should be part of the react UI extension itself

martin-fleck-at replied:
Good point! I added it directly to the React UI extension class.

ivy-lli reviewed (commented):
Hi @martin-fleck-at 

I think this looks very promising, but I think the react stuff should live in its own package. 
So it exports the UI modules, but also the react UI extension, so the inscription package can also import it.
And we should also reuse the components from out https://github.com/axonivy/ui-components repo.

martin-fleck-at commented:
@ivy-lli I updated the PR to extract some React components into its own package but there are not that many overall. So please have a look whether this is what you envisioned. As requested, the components now also reuse some existing components from the `ui-components` repository. Please note that I added the `global.d.ts` file to support having React and VNode classes in the same package. The default is React but if you add the `@jsx svg` annotation, you are expected to use VNode instead.

ivy-lli commented on packages/editor/package.json:
```diff
@@ -19,9 +19,13 @@
   "dependencies": {
     "@axonivy/process-editor-protocol": "~13.1.0-next",
+    "@axonivy/process-editor-view": "~13.1.0-next",
+    "@axonivy/ui-components": "~13.1.0-next.451",
```
this should be a peer dependency like ui-icons

martin-fleck-at replied:
Will change it.

ivy-lli commented on packages/view/package.json:
```diff
@@ -0,0 +1,44 @@
+    "src"
+  ],
+  "dependencies": {
+    "@axonivy/ui-components": "~13.1.0-next.451",
+    "@axonivy/ui-icons": "~13.1.0-next.451"
```
these should be peer dependencies

martin-fleck-at replied:
Will change it.

ivy-lli commented on packages/view/package.json:
```diff
@@ -0,0 +1,44 @@
+    "@types/node": "^22.10.7",
+    "@types/react": "^19.0.8",
+    "deepmerge-ts": "^5.1.0",
+    "jsdom": "^26.0.0",
```
no tests do exist in this package, so this shouldn't be needed

martin-fleck-at replied:
True, I'll remove it.

ivy-lli commented on packages/view/package.json:
```diff
@@ -0,0 +1,44 @@
+  "devDependencies": {
+    "@types/node": "^22.10.7",
+    "@types/react": "^19.0.8",
+    "deepmerge-ts": "^5.1.0",
```
not sure this is needed here either

martin-fleck-at replied:
You are correct!

ivy-lli commented on packages/view/package.json:
```diff
@@ -0,0 +1,44 @@
+  "source": "src/index",
+  "scripts": {
+    "clean": "rimraf lib *.tsbuildinfo",
+    "build": "tsc --build && tscp",
```
that tsc work correctly I think the root tsconfig.json needs to be updated with the reference to this new package. also the editor with uses this needs this reference.
And also the vite configs have references to the src folders of the packages so vite does also watch changes in here.

martin-fleck-at replied:
I don't see the reference for the vite configs but I'll add it to the root tsconfig.json, thank you!

ivy-lli commented on packages/view/package.json:
```diff
@@ -0,0 +1,44 @@
+    "package": "tsc && tscp",
+    "build:clean": "npm run clean && npm run build",
+    "type": "tsc --noEmit",
+    "test": "vitest",
+    "test:ci": "vitest --watch=false"
```
no tests in here so not needed.

ivy-lli commented on packages/editor/src/ui-tools/viewport/ViewportBar.tsx:
```diff
@@ -0,0 +1,24 @@
+const ViewportBar = React.memo(({ triggers, onAction, zoomLevel }: ViewportBarProps) => (
+  <div className='viewport-bar'>
+    <div className='viewport-bar-tools'>
+      {triggers.map(trigger => (
+        <ActionButton key={trigger.id} trigger={trigger} onAction={onAction} />
+      ))}
```
I think I would directly add these here as children.

martin-fleck-at replied:
Yeah, I agree, it is much nicer that way!

ivy-lli commented on packages/editor/src/ui-tools/tool-bar/tool-bar.tsx:
```diff
@@ -86,83 +79,31 @@ export class ToolBar extends GLSPAbstractUIExtension implements IActionHandler,
-    this.toolBarButtonProvider
+  protected render(): React.ReactNode {
+    return (
+      <ToolBarComponent
```
as on the viewport bar, I think it makes not so much sense to pass this buttons as props, I prefer to use the JSX syntax and pass such things as children.

martin-fleck-at replied:
Here, it is a bit different because we do have four different kind of buttons essentially (left, edit, middle, right). I'll see if there is a nicer way of handling this but otherwise if we are just doing children, then the toolbar UI extension needs to specify the whole structure as it has the necessary GLSP context.

ivy-lli reviewed (commented):
Wasn't it the idea to move all the react stuff to the new view package? So e.g. also the toolbar itself? 
In the current state the new package makes not so much sense to me, either move all the react UI extensions to it or then we should not split this up. 
Also I think the UI should use more react primitives e.g. use states for the active button etc.

martin-fleck-at commented:
@ivy-lli I will double check but overall I had a feeling that since we are so closely coupled with GLSP (the action dispatcher, the editor state, the selected elements, etc.), there are not really many components that we can move out since the `ui-components` already exposes a very nice library of elements and we are mostly combining those elements. I'll have another look early next week if we can get some nicer components but so far I had the feeling that most of them are trivial so having their logic in the UI extension is more useful than pushing too much GLSP knowledge into the React components.

ivy-lli reviewed (commented):
Hi @martin-fleck-at 

Sorry I forgot this PR again 🙈 
Thanks for your work, looks great. 

As far as I saw this is currently "missing":
- Palette
- Quickaction bar
However I think we can postpone those things to a later PR. I also started working on a reusable palette component in the ui-components, but only in some free hours...

I'm not sure how we ended last time, did we discuss that with this less components it maybe makes no sense to have a new package and they should maybe moved back to the main editor package?

We also translated the whole editor meanwhile, I think this is the reason again why it has merge conflicts

martin-fleck-at commented:
@ivy-lli I rebased and squashed my commit onto the latest master which should include everything except the quickaction bar for now. Could you please have another look at it?

ivy-lli commented on packages/editor/package.json:
```diff
@@ -19,13 +19,16 @@
   ],
   "dependencies": {
     "@axonivy/process-editor-protocol": "~13.1.0-next",
+    "@axonivy/ui-components": "~13.1.0-next",
```
move to peerDependencies

martin-fleck-at replied:
Done!

ivy-lli reviewed (approved):
Cool.
Meanwhile I also merged the palette, so technically (after the quickaction bar is also done this way) we could also change the menu stuff to do it the reacty way.

martin-fleck-at commented:
> Cool. Meanwhile I also merged the palette, so technically (after the quickaction bar is also done this way) we could also change the menu stuff to do it the reacty way.

Great, that is good news! Is it fine with you to merge this PR first and do that in a follow-up?

ivy-lli commented:
Just saw that no build has run yet... Need to quickly check why.

ivy-lli commented:
Build are running now, sorry for that delay. Hopefully everything is fine

martin-fleck-at commented:
@ivy-lli It seems that one test is failing but I'm not sure I can see which one as I cannot access https://jenkins.ivyteam.io/job/process-editor-client-integration/job/PR-733/1/display/redirect, could you give me the details for it?

ivy-lli commented:
@martin-fleck-at There seems to be one failing test about URL params, maybe just some classes that have changed?

ivy-lli commented:
I think you at least see here which one fails: https://github.com/axonivy/process-editor-client/pull/733/checks?check_run_id=41139931390

ivy-lli commented:
[standalone-chrome] › tests/standalone/url-param/url-param.spec.ts:9:1 › edit mode [standalone-chrome] [standalone-chrome] 

    Error: Timed out 5000ms waiting for expect(locator).toHaveCount(expected)

    Locator: locator('#ivy-tool-bar').locator('.middle-buttons > span')
    Expected: 6
    Received: 0
    Call log:
      - expect.toHaveCount with timeout 5000ms
      - waiting for locator('#ivy-tool-bar').locator('.middle-buttons > span')
        9 × locator resolved to 0 elements
          - unexpected value "0"


       at tests/page-objects/editor/toolbar.ts:103

      101 |     await expect(this.optionsBtn).toBeVisible();
      102 |     await expect(this.toolbar.locator('.edit-buttons')).toBeVisible();
    > 103 |     await expect(this.toolbar.locator('.middle-buttons > span')).toHaveCount(6);
          |                                                                  ^
      104 |   }
      105 |
      106 |   async expectReadonly() {
        at Toolbar.expectEditMode (/home/ubuntu/workspace/editor-client-integration_PR-733/playwright/tests/page-objects/editor/toolbar.ts:103:66)
        at /home/ubuntu/workspace/editor-client-integration_PR-733/playwright/tests/standalone/url-param/url-param.spec.ts:11:3

[trace.zip](https://github.com/user-attachments/files/19905120/trace.zip)

martin-fleck-at commented:
@ivy-lli Thanks! There was indeed a rendering issue with the buttons. I ran the tests locally now, hopefully everything is fine now, could you please re-trigger the build?

martin-fleck-at commented:
@ivy-lli I rebased the PR and re-tested everything locally. For me, all tests run through, not sure why the CI disagreed, could you please re-trigger the tests?

ivy-lli commented:
sorry we had a little problem in our ci dockerfiles
